### DISCUSSION TITLE
Add imports for default values

### DISCUSSION THREAD

**Discussion by St4B** [square/kotlinpoet#1407] [Q&A] [answered]
I am trying to generate a function with the help of KotlinPoet similar to that
```kotlin
fun someFunction(
    dateTime: ZonedDateTime = ZonedDateTime.of(1989,1,23,0,0,0,0, ZoneId.of("UTC"))
) { ... }
```
So I write something like that
```kotlin
FileSpec.builder(packageName = "somePackage", fileName = "someFilename")
    .addFunction(
        FunSpec.builder(name = "someFunction")
            addParameter.(
                 ParameterSpec.builder(
                    name = "dateTime",
                    type = classType
                ).defaultValue("%L", "ZonedDateTime.of(1989,1,23,0,0,0,0, ZoneId.of("UTC"))")
            )
    )
```
The class is generated but compilation fails due to missing import. More specifically, the generated source does not include an import for the `ZoneId`. Obviously, I can add the import manually, but I do not think that this is scalable (similar functions are generated with KSP and some of them can have `ZonedDateTime`, while others may not. Additionally, I have other arguments that are generated like that.  So it does not feel right to add an import for the nested types just of the possibility that they may be existent in the file. Is there any way to handle it?

**Comment by Egorand** [comment, marked as answer]
Imports are automatically added for types (as described in [%T for Types](https://square.github.io/kotlinpoet/#t-for-types)). In your case, "ZonedDateTime" and "ZoneId" are simply parts of an expression, KotlinPoet has no way of knowing that those are actually types. Use the `%T` modifier to resolve this:

```kotlin
defaultValue("%T.of(1989,1,23,0,0,0,0, %T.of(\"UTC\"))", ZonedDateTime::class, ZoneId::class)
```

**Reply by St4B** [reply]
I think that this approach has some limitations. With literals, you are agnostic of the package. I mean that the library that uses KotlinPoet can be agnostic of whether the `ZonedDateTime` is provided from `java.time` or `org.threeten.bp`. With types, this is not possible.

**Reply by Egorand** [reply]
The library can supposedly pass in a `KClass` representing either `java.time.ZonedDateTime` or `org.threeten.bp.ZonedDateTime`, and KotlinPoet can infer the right import from that `KClass`. If you share a specific example I might be able to provide a better suggestion.

**Reply by St4B** [reply]
We have created a KSP processor (you can find it [here](https://github.com/bluegroundltd/fixtures)). We want the KSP processor to be independent of what `ZonedDateTime` object is used. With our current implementation, we are agnostic of the `ZonedDateTime`'s package. The downside is that we need to add the import for `ZoneId` like [that](https://github.com/bluegroundltd/fixtures/blob/d375177db5045bdea7479114f77d500a5375d1a8/fixtures-processor/src/main/kotlin/com/theblueground/fixtures/FixtureBuilderGenerator.kt#L89) which feel ugly and not scalable. 

If we do what you propose, our library will be bound to the package that we have used or we should need to duplicate the things

**Reply by JakeWharton** [reply]
I don't understand either.

If you're generating code based on a user's data class, then use the fully qualified type of the property on that data class.

**Reply by St4B** [reply]
This is what we do. But this is not possible If we had to do something like that
```kotlin
defaultValue("%T.of(1989,1,23,0,0,0,0, %T.of(\"UTC\"))", ZonedDateTime::class, ZoneId::class)
```
because we would have hardcoded the package in our library. My question was if there is a more elegant way to handle with the help of KotlinPoet. I suppose that there is not. Thank you very much for your time :)

**Reply by Egorand** [reply]
The proposed syntax will handle the imports for you, i.e. you won't need to add them manually using `addImport()`. You'd still need to provide the fully-qualified type, either in the form of a `KClass`, or a `ClassName`. If the question is about guessing which version of `ZonedDateTime` is on the classpath and using that one, then KotlinPoet is not really the right tool for that. I presume you can try to resolve a few different flavours of `ZonedDateTime` using the `ClassLoader` and create a `ClassName` instance for the flavour that doesn't fail to resolve.
